[CALCITE-5830] Add ARRAY_INSERT function(enabled in Spark library)

chucheng92 opened the pull request:
# What is the purpose of the change

Add array_insert(x, pos, val) function(enabled in Spark library).

Places val into index pos of array x. Array indices start at 1, or start from the end if index is negative. Index above array size appends the array, or prepends the array if index is negative, with 'null' elements

issue: https://issues.apache.org/jira/browse/CALCITE-5830

Examples:

> SELECT array_insert(array(1, 2, 3, 4), 5, 5); [1,2,3,4,5]

> SELECT array_insert(array(5, 3, 2, 1), -3, 4); [5,4,3,2,1] 

https://spark.apache.org/docs/latest/api/sql/index.html#array_insert

# Brief change log
Add array_insert(x, pos, val) function(enabled in Spark library).

# Verifying this change

SqlOperatorTests#testArrayInsertFunc()

NobiGo commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -4107,6 +4113,97 @@ public static List arrayExcept(List list1, List list2) {
+
+      Object[] newArray = new Object[newArrayLength];
+
+      int posIndex = posInt - 1;
```
Can we use `list.add(pos,element)` ?

chucheng92 replied:
@NobiGo I'm afraid not. 

because our logic here is to move the element on the right of the insertion point to the right, leave the left side unchanged, leaving a space for the target element. If we use `list.add` to insert the left element, there is no problem. When inserting the right element, because we want to reserve a position, it will go out of bounds at this time.

A simple case can explain:
List<String> list = new ArrayList<>();
list.add(0, "A");
list.add(1, "A"); 
list.add(3, "A"); // exception

NobiGo commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -4107,6 +4113,97 @@ public static List arrayExcept(List list1, List list2) {
+    } else {
+      int posIndex = posInt;
+
+      boolean newPosExtendsArrayLeft = -posIndex > baselist.size();
```
How about we use 'posIndex+baselist.size()' >0 or <0? `-posIndex` maybe having a number overflow exception.

chucheng92 commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -4107,6 +4113,97 @@ public static List arrayExcept(List list1, List list2) {
+    boolean usePositivePos = posInt > 0;
+
+    if (usePositivePos) {
+      int newArrayLength = Math.max(baselist.size() + 1, posInt);
```
and here posInt is positive, if posInt < 0, it will from left to insert. 

And the logic of inserting null is the same, as long as `pos` is greater than the length of the array, then calculate the new array length, and fill in the middle with `null`

NobiGo replied:
I get it. Thanks for your explanation. How about we use the `System.arraycopy()` and `Arrays.fill()` to implement this? Others are good to me.

chucheng92 replied:
@NobiGo yes. you are right. `System.arraycopy()` may has better performance when array is large. I have fixed it. PTAL.

chucheng92 commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -4107,6 +4113,97 @@ public static List arrayExcept(List list1, List list2) {
   }
 
+  /** Support the ARRAY_INSERT(array, pos, val) function. */
+  public static @Nullable List arrayInsert(List baselist, Object pos, Object val) {
```
@NobiGo e.g. array_insert(array(1,2,3), 1, 2) calcite will parse and search `arrayInsert[interface java.util.List, int, int]` signature, but it can't match `arrayInsert(List baselist, Integer pos, Object val)`. Java method reflection limits boxing and primitive types can't be matched.

we can other array function. i.e. array_position has same behavior.

chucheng92 replied:
another case, if second pos is a null value. e.g. array_insert(array(1,2,3), cast(null as Integer), 2), then searched method is `arrayInsert[interface java.util.List, Integer, int]`.  We want to use `Integer` to cover `int and Integer`, but it can't be work.

chucheng92 commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -4107,6 +4113,97 @@ public static List arrayExcept(List list1, List list2) {
     return new ArrayList<>(result);
   }
 
+  /** Support the ARRAY_INSERT(array, pos, val) function. */
```
fixed it.

chucheng92 commented:
@NobiGo hi, sorry to ping you. I have solved some comments. If you have time, pls take a look. thanks.

NobiGo commented on site/_docs/reference.md:
```diff
@@ -2658,6 +2658,7 @@ BigQuery's type system uses confusingly different names for types and functions:
 | s | ARRAY_CONTAINS(array, element)                 | Returns true if the *array* contains the *element*
 | s | ARRAY_DISTINCT(array)                          | Removes duplicate values from the *array* that keeps ordering of elements
 | s | ARRAY_EXCEPT(array1, array2)                   | Returns an array of the elements in *array1* but not in *array2*, without duplicates
+| s | ARRAY_INSERT(array, pos, element)              | Places *element* into index *pos* of *array*. Array index start at 1, or start from the end if index is negative. Index above array size appends the array, or prepends the array if index is negative, with 'null' elements.
```
`NULL` value is better?

chucheng92 replied:
yes. fixed

chucheng92 commented:
@NobiGo hi, I have solved all your comments. If you have time, pls help to re-check it. thank you.

NobiGo commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -4107,6 +4113,93 @@ public static List arrayExcept(List list1, List list2) {
+        // special case, if the new position is negative but larger than the current array size
+        // place the new item at start of array, place the current array contents at the end
+        // and fill the newly created array elements in middle with a null
+        int newArrayLength = -posIndex + 1;
```
same question -posIndex maybe having a number overflow exception? Maybe we can use:
`int newArrayLength = Integer.MAX_VALUE - (Integer.MAX_VALUE + posIndex);`
Then the next copy code will be :
`System.arraycopy(baseArray, 0, newArray, newArrayLength - baseArray.length + 1, baseArray.length);`

chucheng92 replied:
thanks for reminding me. however, we can not use
`int newArrayLength = Integer.MAX_VALUE - (Integer.MAX_VALUE + posIndex);`
if posIndex = Integer.MIN_VALUE, it will be 
`int newArrayLength = Integer.MAX_VALUE + 1;` and overflow.

In fact, pos cannot exceed the maximum array limit, we can check in advance just like pos cannot be 0. and -pos will not overflow in subsequent code logic.

I have updated the PR and add overflow test cases. If you are free, PTAL.

mihaibudiu replied:
why not just compute using long values?

chucheng92 replied:
> why not just compute using long values?

thanks for suggestions. the `posIndex` should < Integer.MAX_VALUE - 15, we can simply use Integer to cover it and without overflow. Otherwise, i think it's a good way to check the parameters in advance.

NobiGo commented on core/src/main/java/org/apache/calcite/util/BuiltInMethod.java:
```diff
@@ -658,6 +658,7 @@ public enum BuiltInMethod {
   ARRAYS_OVERLAP(SqlFunctions.class, "arraysOverlap", List.class, List.class),
   ARRAYS_ZIP(SqlFunctions.class, "arraysZip", List.class, List.class),
   SORT_ARRAY(SqlFunctions.class, "sortArray", List.class, boolean.class),
+  ARRAY_INSERT(SqlFunctions.class, "arrayInsert", List.class, Integer.class, Object.class),
```
Put this line to `https://github.com/apache/calcite/pull/3320/files#diff-5207c1dd4af072b3ecbebf9bf87ab0677171bd47309eda809013ef66ae9f77d6R655`. Alphabetic order.

chucheng92 replied:
fixed

NobiGo commented on site/_docs/reference.md:
```diff
@@ -2658,6 +2658,7 @@ BigQuery's type system uses confusingly different names for types and functions:
 | s | ARRAY_CONTAINS(array, element)                 | Returns true if the *array* contains the *element*
 | s | ARRAY_DISTINCT(array)                          | Removes duplicate values from the *array* that keeps ordering of elements
 | s | ARRAY_EXCEPT(array1, array2)                   | Returns an array of the elements in *array1* but not in *array2*, without duplicates
+| s | ARRAY_INSERT(array, pos, element)              | Places *element* into index *pos* of *array*. Array index start at 1, or start from the end if index is negative. Index above array size appends the array, or prepends the array if index is negative, with 'NULL' elements.
```
Usually, We use `` to bracket the keyword. Like: https://github.com/apache/calcite/pull/3320/files#diff-9cda2b29a1b9206e0daa6e6d722eb476575f83eadea1a6e302225afd08d9f0d2R2663

chucheng92 replied:
fixed

JiajunBernoulli commented on testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:
```diff
@@ -5845,6 +5845,84 @@ private static void checkIf(SqlOperatorFixture f) {
+    f1.checkScalar("array_insert(array[1, 3, 4], -2, 2)",
+        "[1, 2, 3, 4]", "INTEGER NOT NULL ARRAY NOT NULL");
+    f1.checkScalar("array_insert(array[2, 3, null, 4], -5, 1)",
+        "[1, null, 2, 3, null, 4]", "INTEGER ARRAY NOT NULL");
```
Why not `[1, 2, 3, null, 4]`?

chucheng92 replied:
same as the question below, i get your point. and I have replied some details in the issue. thanks for reviewing.

JiajunBernoulli commented on testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:
```diff
@@ -5845,6 +5845,84 @@ private static void checkIf(SqlOperatorFixture f) {
+    f1.checkScalar("array_insert(array[2, 3, 4], 1, 1)",
+        "[1, 2, 3, 4]", "INTEGER NOT NULL ARRAY NOT NULL");
+    f1.checkScalar("array_insert(array[1, 3, 4], -2, 2)",
+        "[1, 2, 3, 4]", "INTEGER NOT NULL ARRAY NOT NULL");
```
Why not `[1, 3, 2, 4]`?

chucheng92 replied:
when index is negative, it means 'prepends'. 

array_insert(array[1, 3, 4], -2, 2) => [1, 2, 3, 4]
when we reversed the input array(appends):
array_insert(array[4, 3, 1], 2, 2) => [4, 3, 2, 1]

spark make sure the insert behavior can be equivalent from left or right.

however, [1, 3, 2, 4] is wrong. it's 'appends' behavior.

NobiGo reviewed (approved):
LGTM

liuyongvs commented on core/src/main/java/org/apache/calcite/sql/type/ArrayInsertOperandTypeChecker.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    // op1 is an Integer type
+    final SqlNode op1 = callBinding.operand(1);
```
op1 is an Integer type  +1 great, which aligns with spark

liuyongvs commented on core/src/main/java/org/apache/calcite/sql/type/ArrayInsertOperandTypeChecker.java:
```diff
@@ -0,0 +1,95 @@
+    SqlNode op2 = callBinding.operand(2);
+    RelDataType op2Type = SqlTypeUtil.deriveType(callBinding, op2);
+    if (op2Type.getSqlTypeName() != op0ComponentType.getSqlTypeName()) {
+      if (throwOnFailure) {
```
op2Type.getSqlTypeName() != op0ComponentType.getSqlTypeName())
it may not enougn, when the type is also complex type, like array/row/spark. for example row(int) and row(int , bigint)

liuyongvs replied:
op2Type.getSqlTypeName() != op0ComponentType.getSqlTypeName())
it may not enougn, when the type is also complex type, like array/row/spark. for example row(int) and row(int , bigint)

liuyongvs commented:
looks great, only left one comments, which should be changed @chucheng92

chucheng92 commented:
> looks great, only left one comments, which should be changed @chucheng92

thanks Lau, you are right. here we need also using leastRestrictive type just like other array function. I have updated the PR and added the complex type cases. If you have time, pls check it. thanks.

chucheng92 commented:
Thank you all for patient reviewing, all comments are resolved. @NobiGo hi, xiong, If you help time, could you help to merge it?
